ledger/verify_block.go

```diff
@@ -23,124 +23,395 @@ import (
+				err,
+			)
+		}
+		signature = h.Signature
+		hotVkey = h.Body.OpCertHotVkey
+		kesPeriod = uint64(h.Body.OpCertKesPeriod)
+	case BlockTypeAlonzo:
+		h := specificHeader.(*alonzo.AlonzoBlockHeader)
+		bodyCbor, err = cbor.Encode(h.Body)
+		if err != nil {
+			return false, "", 0, 0, fmt.Errorf(
+				"VerifyBlock: failed to encode Alonzo header body for KES, %w",
+				err,
+			)
+		}
+		signature = h.Signature
+		hotVkey = h.Body.OpCertHotVkey
+		kesPeriod = uint64(h.Body.OpCertKesPeriod)
+	case BlockTypeBabbage:
+		h := specificHeader.(*babbage.BabbageBlockHeader)
+		bodyCbor, err = cbor.Encode(h.Body)
+		if err != nil {
+			return false, "", 0, 0, fmt.Errorf(
+				"VerifyBlock: failed to encode Babbage header body for KES, %w",
+				err,
+			)
+		}
+		signature = h.Signature
+		hotVkey = h.Body.OpCert.HotVkey
+		kesPeriod = uint64(h.Body.OpCert.KesPeriod)
+	case BlockTypeConway:
+		h := specificHeader.(*conway.ConwayBlockHeader)
+		bodyCbor, err = cbor.Encode(h.Body)
+		if err != nil {
+			return false, "", 0, 0, fmt.Errorf(
+				"VerifyBlock: failed to encode Conway header body for KES, %w",
+				err,
+			)
+		}
+		signature = h.Signature
+		hotVkey = h.Body.OpCert.HotVkey
+		kesPeriod = uint64(h.Body.OpCert.KesPeriod)
+	default:
 		return false, "", 0, 0, fmt.Errorf(
-			"VerifyBlock: VerifyBlockBody error, %v",
-			isBodyValidError.Error(),
+			"VerifyBlock: unsupported block type for KES %d",
+			blockType,
 		)
 	}
-	isValid = isKesValid && isVrfValid && isBodyValid
-	vrfHex = hex.EncodeToString(vrfBytes)
-	blockNo := header.Body.BlockNumber
-	slotNo := header.Body.Slot
-	return isValid, vrfHex, blockNo, slotNo, nil
-}
 
-func ExtractBlockData(
-	bodyHex string,
-) ([]UTXOOutput, []RegisCert, []DeRegisCert, error) {
-	rawDataBytes, rawDataBytesError := hex.DecodeString(bodyHex)
-	if rawDataBytesError != nil {
-		return nil, nil, nil, fmt.Errorf(
-			"ExtractBlockData: bodyHex decode error, %v",
-			rawDataBytesError.Error(),
-		)
-	}
-	var txsRaw [][]string
-	_, err := cbor.Decode(rawDataBytes, &txsRaw)
+	kesValid, err = VerifyKesComponents(
+		bodyCbor,
+		signature,
+		hotVkey,
+		kesPeriod,
+		slot,
+		slotsPerKesPeriod,
+	)
 	if err != nil {
-		return nil, nil, nil, fmt.Errorf(
-			"ExtractBlockData: txsRaw decode error, %v",
+		return false, "", 0, 0, fmt.Errorf(
+			"VerifyBlock: KES verification error, %v",
 			err.Error(),
 		)
 	}
-	txBodies, txBodiesError := GetTxBodies(txsRaw)
-	if txBodiesError != nil {
-		return nil, nil, nil, fmt.Errorf(
-			"ExtractBlockData: GetTxBodies error, %v",
-			txBodiesError.Error(),
-		)
-	}
-	uTXOOutput, regisCerts, deRegisCerts, getBlockOutputError := GetBlockOutput(
-		txBodies,
-	)
-	if getBlockOutputError != nil {
-		return nil, nil, nil, fmt.Errorf(
-			"ExtractBlockData: GetBlockOutput error, %v",
-			getBlockOutputError.Error(),
-		)
+
+	// Verify block body hash
+	expectedBodyHash := block.BlockBodyHash()
+	isBodyValid := true
+	if block.Era().Name != "Byron" {
```

Comment:
This should be a comparison to a pre-defined value for Byron rather than a string. Something like this:

```
if block.Era() != byron.EraByron {
```